### PR TITLE
fix: JSON Schema for variable properties

### LINKED ISSUE #367: Variable Property Name doesn't seem to be correctly implemented
If I try to parse

```apib
FORMAT: 1A

# Beehive API

## Honey [/honey]

### Retrieve [GET]

+ Response 200 (application/json)

    + Attributes

        + todosById
            + *8*
                + text: Prepare Dinner (string, required)
```

with `drafter@3.0.0`, I get

#### JSON

```json
{
  "todosById": {
    "8": {
      "text": "Prepare Dinner"
    }
  }
}
```

#### JSON Schema

```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "todosById": {
      "$ref": "#/definitions/8"
    }
  },
  "definitions": {
    "8": {
      "type": "object",
      "properties": {
        "text": {
          "type": "string"
        }
      },
      "required": [
        "text"
      ]
    }
  }
}
```

I think the schema is incorrect, because the property has [Variable Property Name](https://github.com/apiaryio/mson/blob/master/MSON%20Specification.md#322-variable-property-name), but the schema looks like it forces you to have a property called `8` exactly. In my opinion this is a bug.

--------------------

This issue was filed as an outcome of investigation of https://github.com/apiaryio/dredd/issues/534

### PR DESCRIPTION
**Pull request by pksunkara**
Fixes #367. But I discovered a new bug #413. No, I don't think it should be fixed in this PR because it is completely different.

### CODE REVIEW

**Comment by pksunkara** [comment]
I have rebased this on master now.

File: src/refract/JSONSchemaVisitor.cc [hunk 1]
```diff
@@ -222,23 +222,39 @@ namespace refract
         }
     }
 
+    ObjectElement* JSONSchemaVisitor::definitionFromVariableProperty(JSONSchemaVisitor& renderer)
+    {
+        ObjectElement* definition = new ObjectElement();
+
+        definition->push_back(new MemberElement("type", new StringElement("object", IElement::rCompact), IElement::rCompact));
```
Comment (w-vi) [inline src/refract/JSONSchemaVisitor.cc]: Can these super long lines be somehow indented?

Comment (pksunkara) [inline src/refract/JSONSchemaVisitor.cc]: I don't think it's that long. It's only 126,

Comment (w-vi) [inline src/refract/JSONSchemaVisitor.cc]: And I do ... so this brings again the bloody style issue, what is our column limit? Is it 80, 120, 280? According to the clang format it is 120, https://github.com/apiaryio/snowcrash/pull/380/files#diff-69a68f7ce1ea3687a38d8e4705e6fa48R25

Comment (pksunkara) [inline src/refract/JSONSchemaVisitor.cc]: I can fix this later once you finish the review.

**Comment by w-vi** [comment]
Fixture https://github.com/apiaryio/drafter/blob/master/test/fixtures/schema/variable-property2.apib

produces following schema 
```
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "item": {
      "allOf": [
        {
          "$ref": "#/definitions/link"
        },
        {
          "properties": {
            "id": {
              "type": "object",
              "properties": {
                "name": {
                  "type": "string"
                },
                "email": {
                  "type": "string"
                }
              }
            }
          }
        }
      ]
    }
  },
  "definitions": {
    "link": {
      "type": "object",
      "patternProperties": {
        "": {
          "type": "object",
          "properties": {
            "href": {
              "type": "string"
            }
          }
        }
      }
    }
  }
}
```

but I am not sure if this is correct as following JSON seems to validate against it event though iot seems invalid to me. Maybe we have some fundamental flaw in our design.

```
{
 "l": {
      "h": "a URI"
    },
    "i": {
      "n": "",
      "e": ""
    }
}
```

**Comment by pksunkara** [comment]
Maybe the validator is wrong? 😄. I will look into it.

**Comment by pksunkara** [comment]
Actually, we do not do any required properties here and we are not restricting random properties. I guess @honzajavorek can test this for us in dredd since this issue came from him originally.

**Comment by w-vi** [comment]
You are right, otherwise it LGTM.

**Comment by pksunkara** [comment]
@w-vi Will you be merging this?

**Comment by w-vi** [comment]
I have another example which I am not sure what the result should be.

Take the https://github.com/apiaryio/drafter/blob/master/test/fixtures/schema/variable-property.apib and make the `*self*` required.  The schema is definitely not correct. 

```

{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "_links": {
      "$ref": "#/definitions/self",
      "required": [
        "self"
      ]
    }
  },
  "definitions": {
    "self": {
      "type": "object",
      "patternProperties": {
        "": {
          "type": "object",
          "properties": {
            "href": {
              "type": "string"
            }
          }
        }
      }
    }
  }
}
```

**Comment by pksunkara** [comment]
Well, the schema is correct. MSON means that a variable property should be present. And JSON Schema uses `self` as a sample. So, It's kinda correct.

The issue is conceptually, it doesn't mean much in MSON. Since both required and variable property are targetted at difference scenarios.

Maybe, later we can add a warning saying that variable property can't be required or optional.

**Comment by w-vi** [comment]
@pksunkara please rebase so it can be merged.

I think we are still not handlingg  the schema correctly, but at least now we let slip something rather then restricting it.

**Comment by w-vi** [comment]
Oh and please indent the long lines.

**Comment by pksunkara** [comment]
Done and done.

**Pull request merged**